src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+                        {
+                            // Populate our map using the initial data.
+                            _resourceByName.Clear();
+
+                            // TODO send a "clear" event via outgoing channels, in case consumers have extra items to be removed
+
+                            foreach (var resource in response.InitialData.Resources)
+                            {
+                                // Add to map.
+                                var viewModel = resource.ToViewModel();
+                                _resourceByName[resource.Name] = viewModel;
+
+                                // Send this update to any subscribers too.
+                                changes ??= [];
+                                changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                            }
+                        }
+                        else if (response.KindCase == WatchResourcesUpdate.KindOneofCase.Changes)
+                        {
+                            // Apply changes to the model.
+                            foreach (var change in response.Changes.Value)
+                            {
+                                changes ??= [];
+
+                                if (change.KindCase == WatchResourcesChange.KindOneofCase.Upsert)
+                                {
+                                    // Upsert (i.e. add or replace)
+                                    var viewModel = change.Upsert.ToViewModel();
+                                    _resourceByName[change.Upsert.Name] = viewModel;
+                                    changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                                }
+                                else if (change.KindCase == WatchResourcesChange.KindOneofCase.Delete)
+                                {
+                                    // Remove
+                                    if (_resourceByName.Remove(change.Delete.ResourceName, out var removed))
+                                    {
+                                        changes.Add(new(ResourceViewModelChangeType.Delete, removed));
+                                    }
+                                    else
+                                    {
+                                        Debug.Fail("Attempt to remove an unknown resource view model.");
+                                    }
+                                }
+                                else
+                                {
+                                    throw new FormatException($"Unexpected {nameof(WatchResourcesChange)} kind: {change.KindCase}");
+                                }
+                            }
+                        }
+                        else
+                        {
+                            throw new FormatException($"Unexpected {nameof(WatchResourcesUpdate)} kind: {response.KindCase}");
```

Comment:
nit: FormatException isn't the exception I'd expect here. I would expect ArgumentOutOfRangeException or InvalidOperationException

Reply:
`ArgumentOutOfRangeException` is for arguments, and this is not an argument.

`InvalidOperationException` implies that the requested operation is not applicable given the current state of the object, which does not describe this situation.

The exception here is that the received data is not of the expected format. `FormatException` is used in several scenarios, commonly around parsing, which is what we're doing here.

We could create a custom exception type, but I don't think it's necessary.